### PR TITLE
Expose matchLink method to override

### PR DESCRIPTION
**Pull request by gshahbazian**
We can override this with our own linkify logic.

### CODE REVIEW

File: build/squire-raw.js [hunk 1]
```diff
@@ -4215,7 +4215,7 @@ proto.insertImage = function ( src, attributes ) {
     return img;
 };
 
-var linkRegExp = /\b((?:(?:ht|f)tps?:\/\/|www\d{0,3}[.]|[a-z0-9.\-]+[.][a-z]{2,}\/)(?:[^\s()<>]+|\([^\s()<>]+\))+(?:\((?:[^\s()<>]+|(?:\([^\s()<>]+\)))*\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’]))|([\w\-.%+]+@(?:[\w\-]+\.)+[A-Z]{2,}\b)/i;
+var linkRegExp = /\b(?:(?:ht|f)tps?:\/\/)?(?:[\da-z\.-@]+)\.(?:[a-z\.]{2,6})(?:[\/\w\.-?#]*)*/i;
```
Comment (ibash) [inline build/squire-raw.js]: Where is this regex from? I think we have a few different url regexes in the code base

Comment (gshahbazian) [reply build/squire-raw.js]: I googled around and altered it to also catch emails

**Comment by ibash** [comment]
@gshahbazian this needs a test

**Comment by ibash** [comment]
Also take a look at app/service/linkify.cjsx in our code base

**Comment by gvohra** [comment]
Re: test, we can use sandboxed_render.cjsx to look for CSS

Sent via Superhuman ( https://sprh.mn/?vip=gaurav@superhuman.com )

**Comment by gvohra** [comment]
sandboxed_render_test.cjsx*

Sent via Superhuman ( https://sprh.mn/?vip=gaurav@superhuman.com )

**Comment by gshahbazian** [comment]
@gvohra I think @ibash means a squire test

**Comment by ibash** [comment]
@gshahbazian this is to fix what issue?

**Comment by gshahbazian** [comment]
@ibash https://github.com/superhuman/superhuman/issues/4639

**Comment by ibash** [comment]
Testing it out — I think this is a bit too liberal in what it considers a link:
screen shot 2017-08-15 at 1 56 38 pm

In our linkify service we check tlds explicitly -- I think if we're going to do this we should be checking tlds.

**Comment by gshahbazian** [comment]
@ibash I agree. Trying out this one https://gist.github.com/gruber/8891611

**Comment by ibash** [comment]
@gshahbazian another approach could be to use our linkify regex directly -- and instead of fixing this at the squire level, overriding the space/enter/paste handlers in superhuman and checking it then. We do this in `app/editor/` I'd be happy to show you this.

**Comment by gshahbazian** [comment]
@ibash updated. This version no longer linkifies "over.You" but will linkify "http://over.You"

File: source/Editor.js [hunk 2]
```diff
@@ -1702,6 +1702,7 @@ var addLinks = function ( frag, root, self ) {
         data = node.data;
         parent = node.parentNode;
         while ( match = linkRegExp.exec( data ) ) {
+            console.log(match);
```
Comment (ibash) [inline source/Editor.js]: Please remove :)

**Comment by gshahbazian** [comment]
@ibash test added

File: source/Editor.js [hunk 3]
```diff
@@ -1688,7 +1688,7 @@ proto.insertImage = function ( src, attributes ) {
...
...tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|yu|za|zm|zw)\/)(?:[^\s()<>{}\[\]]+|\([^\s()]*?\([^\s()]+\)[^\s()]*?\)|\([^\s]+?\))+(?:\([^\s()]*?\([^\s()]+\)[^\s()]*?\)|\([^\s]+?\)|[^\s`!()\[\]{};:'".,<>?«»“”‘’])|(?:[a-z0-9]+(?:[.\-][a-z0-9]+)*[.](?:com|net|org|edu|gov|mil|aero|asia|biz|cat|coop|info|int|jobs|mobi|museum|name|post|pro|tel|travel|xxx|ac|ad|ae|af|ag|ai|al|am|an|ao|aq|ar|as|at|au|aw|ax|az|ba|bb|bd|be|bf|bg|bh|bi|bj|bm|bn|bo|br|bs|bt|bv|bw|by|bz|ca|cc|cd|cf|cg|ch|ci|ck|cl|cm|cn|co|cr|cs|cu|cv|cx|cy|cz|dd|de|dj|dk|dm|do|dz|ec|ee|eg|eh|er|es|et|eu|fi|fj|fk|fm|fo|fr|ga|gb|gd|ge|gf|gg|gh|gi|gl|gm|gn|gp|gq|gr|gs|gt|gu|gw|gy|hk|hm|hn|hr|ht|hu|id|ie|il|im|in|io|iq|ir|is|it|je|jm|jo|jp|ke|kg|kh|ki|km|kn|kp|kr|kw|ky|kz|la|lb|lc|li|lk|lr|ls|lt|lu|lv|ly|ma|mc|md|me|mg|mh|mk|ml|mm|mn|mo|mp|mq|mr|ms|mt|mu|mv|mw|mx|my|mz|na|nc|ne|nf|ng|ni|nl|no|np|nr|nu|nz|om|pa|pe|pf|pg|ph|pk|pl|pm|pn|pr|ps|pt|pw|py|qa|re|ro|rs|ru|rw|sa|sb|sc|sd|se|sg|sh|si|sj|Ja|sk|sl|sm|sn|so|sr|ss|st|su|sv|sx|sy|sz|tc|td|tf|tg|th|tj|tk|tl|tm|tn|to|tp|tr|tt|tv|tw|tz|ua|ug|uk|us|uy|uz|va|vc|ve|vg|vi|vn|vu|wf|ws|ye|yt|yu|za|zm|zw)\b\/?(?!@)))|([\w\-.%+]+@(?:[\w\-]+\.)+[A-Z]{2,}\b)/i
```
Comment (ibash) [inline source/Editor.js]: This loses the ability to match `ftp://` :p probably nbd

Where is this regex from? `[a-z0-9.\-]+[.]` the second `[.]` is interesting - why is that in a character class by itself?

Comment (gshahbazian) [reply source/Editor.js]: https://gist.github.com/gruber/8891611

Comment (ibash) [reply source/Editor.js]: Cool -- was this modified at all?

**Comment by ibash** [comment]
Tested this on the text in linkify_test.cjsx -- it misses a few cases we catch, but those are the odd edge cases.

I'm a little bit wary of catastrophic backtracking, we've had to deal with that before because of our linkify regex, and it's not a fun thing to debug (the symptoms are that the cpu goes to 100% and chrome locks up - previously we had to debug in person because certain strings would trigger it and we couldn't reproduce locally).

So, if you think the regex is well-vetted (looks like it may be?) then I think this is okay

screencapture-file-users-islam-projects-squire-demo-html-1502840300593

**Comment by ibash** [comment]
@gshahbazian one more thing is that I think this will regress cases where we catch a link that squire does not, when generating html to send we do:
```
  generateForOutgoingMessage: (readStatusPixel) =>
    # What this does:
    # 1. Linkify urls
    # 2. Add inline images
    # 3. Wraps the draft body with signatures, a read status pixel, and quoted content
    # 4. Cleans superhuman media links

    dom = @_rehydrateToDOM(@draft.getBody())
    Linkify.element(dom)
```

If squire has already linkified something I think we'll miss it when we call Linkify.element

**Comment by ConradIrwin** [comment]
@ibash @gshahbazian rxxr2 thinks this is probably OK (but it doesn't support the negative lookahead @-sign, so I had to modify the regex to get it to check).

That said I would also like consistency instead of multiple ways of doing the same thing.

File: source/Editor.js (hunk 3 above)
Comment (gshahbazian) [reply source/Editor.js]: @ibash slightly, removed some pearl only regex stuff, then added the second capture group that squire already had to find emails.

**Comment by gshahbazian** [comment]
@ibash @ConradIrwin removed the regex logic, exposed an override point.

**Comment by ibash** [comment]
shipit

**Pull request merged**